xarray/backends/zarr.py

```diff
@@ -602,6 +602,8 @@ class ZarrStore(AbstractWritableDataStore):
 
     __slots__ = (
         "_append_dim",
+        "_cache",
+        "_cache_array_keys",
```

Comment:
We also loop over `.arrays()` twice. I think we should cache the output of that instead, and construct `arrays_keys` from the cached `arrays()` result.

Reply:
good idea!